### PR TITLE
[tiff] Update to 4.3.0, improvements

### LINKED ISSUE #14849: [tiff] Installed pkg-config (.pc) debug files specify non-existent includedir
****

**Is your feature request related to a problem? Please describe.**

I'm trying to package a new port. The port uses `pkg-config` to find dependencies. It also depends on `tiff`. So in my cmake file I have

```
find_package(PkgConfig REQUIRED)
pkg_check_modules(libtiff-4 IMPORTED_TARGET REQUIRED libtiff-4)
```

On x64-windows `vcpkg install <my port>` fails with the following error:

```
  CMake Error in src/sail-codecs/src/tiff/CMakeLists.txt:
  Imported target "PkgConfig::libtiff-4" includes non-existent path

    "D:/projects/git.vcpkg/packages/tiff_x64-windows/debug/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

**Proposed solution**
Fix libtiff to make sure all the paths specified in its .pc files really exist

**Describe alternatives you've considered**
Created `D:/projects/git.vcpkg/packages/tiff_x64-windows/debug/include` by hand

**Additional context**
-

### LINKED ISSUE #18011: [tiff] update to 4.3.0
Library name:
LibTiff
New version number:
4.3.0
Other information that may be useful (release notes, etc...)

At the moment, the latest stable release of LibTiff is 4.3.0 but this request is to update it to the latest. Information on 4.3.0 can be found here: 
http://www.simplesystems.org/libtiff/v4.3.0.html

The latest version of LibTiff contains fixes for the following Security Vulnerabilities:
https://nvd.nist.gov/vuln/detail/CVE-2020-35521
https://nvd.nist.gov/vuln/detail/CVE-2020-35522
https://nvd.nist.gov/vuln/detail/CVE-2020-35523
https://nvd.nist.gov/vuln/detail/CVE-2020-35524

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?  
  - [x] Updates tiff to 4.3.0. Fixes #18011.
  - [x] Revised feature interface. C++ lib tiffcxx is opt-in now. Former "tool" is "tools" now.
  - [x] Fixes pc and cmake config files. Fixes #14849.
  - [x] Migrate to vcpkg-cmake ports

- #### Important change in libtiff 4.3.0
  > Use of "int8", "uint8", "int16", "uint16", "int32", "uint32", "int64", "uint64" typedefs is now deprecated. libtiff code and headers no longer use them, and use their C99 standard equivalents (with "_t" suffix). Those typedefs are still available, with deprecation warnings, but external code is strongly encouraged to use the corresponding C99 stdint.h types. The deprecated types might be removed in a future release.

  (From https://libtiff.gitlab.io/libtiff/v4.3.0.html#libtiff) 
  This affects at least port selene which treated these warnings as errors.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  I am still working on this PR.

### CODE REVIEW

File: ports/tiff/cmakelists.patch [hunk 1]
```diff
@@ -1,84 +1,58 @@
...
 --- a/libtiff/CMakeLists.txt
 +++ b/libtiff/CMakeLists.txt
-@@ -143,7 +143,7 @@ install(FILES ${tiff_HEADERS} ${nodist_tiff_HEADERS}
-         DESTINATION "${CMAKE_INSTALL_FULL_INCLUDEDIR}")
+@@ -161,7 +161,7 @@ if(CXX_SUPPORT)
+   set(tiffxx_HEADERS
+           tiffio.hxx)
+ 
+-  add_library(tiffxx)
++  add_library(tiffxx STATIC)
```
Comment (cenit) [inline ports/tiff/cmakelists.patch:42]: is it necessary?

Comment (dg0yt) [reply ports/tiff/cmakelists.patch:42]: The `static` was there in the original patch. I removed it, and it turned out that it is missing proper export/import markup for windows DLLs.
I don't care much about tiffxx in particular, and I made it an optional feature in the commit which will follow now.

Comment (cenit) [reply ports/tiff/cmakelists.patch:42]: thanks for the clarification

**Comment by dg0yt** [comment]
Manually tested, covering the cmake wrapper:
`tiff[cxx,jpeg,lzma,tools,webp,zip,zstd]:x64-linux + libgeotiff:x64-linux`

File: ports/tiff/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
+{
+  "name": "tiff",
+  "version-string": "4.3.0",
```
Comment (JonLiu1993) [inline ports/tiff/vcpkg.json:3]: @dg0yt ,We should replace "version-string" to "version"

File: versions/t-/tiff.json [hunk 3]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "bdb1bd990e5d6fd111baca5f2c8e1513c4a66068",
+      "version-string": "4.3.0",
```
Comment (JonLiu1993) [inline versions/t-/tiff.json:5]: We should replace "version-string" to "version"

File: ports/selene/portfile.cmake [hunk 4]
```diff
@@ -8,6 +8,7 @@ vcpkg_from_github(
     HEAD_REF master
     PATCHES
         disable_x86_intrinsics_on_arm.patch
+        tiff-deprecated-typedefs.patch
 )
 
 vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS "opencv" SELENE_USE_OPENCV)
```
Comment (JonLiu1993) [inline ports/selene/portfile.cmake:14]: We should add "FEATURES" to the function vcpkg_check_features 

vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS 
    FEATURES    
    "opencv"     SELENE_USE_OPENCV)

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

File: ports/selene/portfile.cmake (hunk 4 above)
Comment (dg0yt) [reply ports/selene/portfile.cmake]: I modernized the whole port now.

File: ports/selene/tiff-deprecated-typedefs.patch [hunk 5]
```diff
@@ -0,0 +1,14 @@
+diff --git a/selene/CMakeLists.txt b/selene/CMakeLists.txt
+index ab2760c..d1e1e6c 100644
+--- a/selene/CMakeLists.txt
++++ b/selene/CMakeLists.txt
+@@ -229,6 +229,9 @@ if(TIFF_FOUND)
+             )
+ 
+     target_compile_options(selene_img_io_tiff PRIVATE ${SELENE_COMPILER_OPTIONS} ${SELENE_IMG_COMPILER_OPTIONS})
++    if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
```
Comment (JackBoosY) [inline ports/selene/tiff-deprecated-typedefs.patch:9]: Better:
```cmake
    if (MSVC)
```

**Review by JackBoosY — APPROVED** [review]

File: ports/tiff/portfile.cmake [hunk 6]
```diff
@@ -1,84 +1,101 @@
...
 file(REMOVE_RECURSE
-    ${CURRENT_PACKAGES_DIR}/debug/include
-    ${CURRENT_PACKAGES_DIR}/debug/share
-    ${CURRENT_PACKAGES_DIR}/share
+    "${CURRENT_PACKAGES_DIR}/debug/include"
+    "${CURRENT_PACKAGES_DIR}/debug/share"
+    "${CURRENT_PACKAGES_DIR}/share"
 )
 
+configure_file("${CMAKE_CURRENT_LIST_DIR}/vcpkg-cmake-wrapper.cmake" "${CURRENT_PACKAGES_DIR}/share/${PORT}/vcpkg-cmake-wrapper.cmake" @ONLY)
```
Comment (cenit) [inline ports/tiff/portfile.cmake:72]: i think it's better to rename the wrapper and add a .in to the original non-configured one

**Review by cenit — APPROVED** [review]

**Comment by JonLiu1993** [comment]
All features are tested successfully in the following triplets:
- x86-windows
- x64-windows
- x64-windows-static

**Review by JonLiu1993 — APPROVED** [review]

**Review by ras0219-msft — APPROVED** [review]
LGTM, thanks for the update!

**Comment by strega-nil-ms** [comment]
LGTM, thanks @dg0yt :)

**Comment by AbdulAbCellera** [comment]
Hi, I have a question about CXX feature support on `arm`. 

I want to compile with the CXX feature for Apple M1 (arm64). So I did the following:

- I removed `"supports" : !arm`
- I linked my targets to TIFF::CXX
- I compiled

These changes do not seem to have any issues. 

The question is what is the best approach to get this support natively? 
Do I have to link to another target? 
or post a PR for arm64 support of CXX feature?

**Comment by dg0yt** [comment]
I guess this particular arm restriction is obsolete. The next PR can remove the `supports` line from this feature. In the mean time you may also install `tiff[cxx]` with `--allow-unsupported`.

**Comment by AbdulAbCellera** [comment]
Is there a way to set `--allow-unsupported` in manifest mode?

**Comment by dg0yt** [comment]
> Is there a way to set `--allow-unsupported` in manifest mode?

Yes, just pass is through with `CMAKE_INSTALL_OPTIONS`:
https://github.com/microsoft/vcpkg/blob/master/docs/users/buildsystems/cmake-integration.md#vcpkg_install_options

**Comment by AbdulAbCellera** [comment]
I think you meant `VCPKG_INSTALL_OPTIONS` in stead of `CMAKE_INSTALL_OPTIONS` :)

**Pull request merged**